src/appleseed/foundation/utility/unzipper.cpp

```diff
@@ -0,0 +1,197 @@
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2017 Gleb Mishchenko, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "unzipper.h"
+
+// appleseed.foundation headers.
+#include "foundation/core/exceptions/exception.h"
+#include "foundation/utility/minizip/unzip.h"
+
+// Boost headers.
+#include "boost/filesystem.hpp"
+
+// Standard headers.
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <sstream>
+
+using namespace std;
+namespace bf = boost::filesystem;
+
+namespace foundation
+{
+
+const size_t BUFFER_SIZE = 4096;
+
+bool is_zip_entry_directory(const string& dirname) 
+{
+    // used own implementation of is_zip_entry_directory instead of boost implementation
+    // because this directory is not in filesystem, but in zipfile
+    return dirname[dirname.size() - 1] == '/';
+}
+
+void open_current_file(unzFile& zip_file) 
+{
+    const int err = unzOpenCurrentFile(zip_file);
+    if (err != UNZ_OK) 
+        throw UnzipException("Can't open file inside zip: ", err);
+}
+
+int read_chunk(unzFile& zip_file, char* buffer, const int chunk_size) 
+{
+    const int err = unzReadCurrentFile(zip_file, buffer, chunk_size);
+
+    if (err == UNZ_ERRNO)
+        throw UnzipException("IO error while reading from zip");
+    if (err < 0)
+        throw UnzipException("zLib error while decompressing file: ", err);
+
+    return err;
+}
+
+void close_current_file(unzFile& zip_file) 
+{
+    const int err = unzCloseCurrentFile(zip_file);
+
+    if (err == UNZ_CRCERROR)
+        throw UnzipException("CRC32 is not good");
+}
+
+string read_filename(unzFile& zip_file) 
+{
+    unz_file_info zip_file_info;
+    unzGetCurrentFileInfo(zip_file, &zip_file_info, NULL, 0, NULL, 0, NULL, 0);
+
+    char* filename = new char[zip_file_info.size_filename + 1];
```

Comment:
We can do even better I think, and in particular we can make this code exception-safe (which it isn't currently: if `unzGetCurrentFileInfo()` throws, memory is leaked).

With C++11 and beyond we probably could do

    string inzip_filename;
    inzip_filename.resize(zip_file_info.size_filename);
    unzGetCurrentFileInfo(zip_file, &zip_file_info, &inzip_filename[0], zip_file_info.size_filename, NULL, 0, NULL, 0);

because `std::string` is guaranteed to use contiguous storage.

Unfortunately, until the next release we are still stuck to C++03, where this would be better:

    vector filename(zip_file_info.size_filename + 1);
    unzGetCurrentFileInfo(zip_file, &zip_file_info, &filename[0], filename.size(), NULL, 0, NULL, 0);
    const string inzip_filename(&filename[0]);

To be tested (especially regarding the +1 for `\0` etc.)